### DISCUSSION TITLE
Is it required that the two models of the input "buildbematrix" module are closed?

### DISCUSSION THREAD

**Discussion by zhifeiji-han** [SCIInstitute/SCIRun#2300] [General]
I have a few doubts when learning the example program “torso-tank-bem.srn5”.
1.Is the imported cage.fld data epicardial geometry? Why is the shape a cylinder? 
![image](https://user-images.githubusercontent.com/70640243/122764330-f415cf80-d2d1-11eb-974a-1fd7d693ca11.png)
2.The dimension of the transfer matrix is 771 * 602,  771 represents the number of torso surface nodes and 602 represents the number of epicardial nodes(I think). When calculating the transfer matrix, Why is epicardium mapped to the whole torso instead of epicardium mapped to the sensor position on the surface of the torso?
3.Is it required that the two models of the input "buildbematrix" module are closed?
4.How to get  "tank. FLD"_ measured_ Only. fld " from "tank.fld"? Is it possible to specify or program it?

**Comment by jab0707** [comment]
1) The cage geometry represents a pericardiac cage that records potentials from 2-3 mm off of the heart surface, so it is not a true "epicardial" measure per say but close to it.
2) 771 is the number of nodes in the full closed torso surface mesh. If you want to just map to the sensor locations then you would extract the rows corresponding to those node indicies. The reason you would calculate it on the full 771 torso is because that torso is a closed surface, which getting to point 
3) it does need to be closed/it should. The BEM method is based in part on the Barr paper (https://ieeexplore.ieee.org/document/4122629) which is formulated for closed surfaces. 
4) I am not sure what yoiu are asking for here. Do you mean how do you only get the measurement nodes?

**Comment by jessdtate** [comment]
This BEM implementation creates a transfer function between close surfaces.  All the points on both.  

If you want a transfer matrix to the measurement nodes, or the sensor positions, just use buildmappingmatrix to create a mapping from the 771 torso to the measurement points.  Make sure that both fields have a linear basis before making the mapping matrix using convertfieldbasis.  once you have the mapping matrix, just premultiply it with the BEM matrix and you will a matrix that is between the cage and the measurement points.

Also, the shape is a cylinder because that was the shape of the cage used in the experiment.

**Comment by zhifeiji-han** [comment]
1.About question 4 last time, I mean how to get the measurement node?(as you mentioned)
2.As for the calculation and use of ECG  transfer matrix, the calculation process is like this (I think).
Specifically, the transfer matrix A1 corresponding to the projection from the closed epicardium to the closed torso is obtained (the number of rows of A1 is the number of torso nodes, which is far more than the number of sensors), and then the number of rows for sensor array position nodes is extracted from the  transfer matrix (A1). The new  transfer matrix A2 is obtained, which can be used for ECG inverse solution (the number of rows of the new  transfer matrix A2 corresponds to the number of sensors). 
Is that right?

**Comment by zhifeiji-han** [comment]
@jessdtate 
According to your suggestion, I have established the transfer matrix from epicardium to sensor（I use "tank_ measured_ Only. fld "）.
However, I still have no idea to obtain the sensor array node position on the torso surface. Is there any way to do this？

**Comment by jessdtate** [comment]
The measurement nodes are just another geometry that was made to reflect the recording locations.  In these data, the electrode locations were recorded with a digitizer and the triangulation was made manually.   The triangulation helps with mapping and visualization.  

It sounds like you ware describing the process for making a transfer matrix from the heart to recording locations accurately.  

One more point about the full 771 torso surface: it more accurately represents the experimental setup than a closed geometry with only the recording electrodes.  That's another reason we do it this way. 

There are many ways to obtain sensor positions in a given setup, and it mostly depends on the setting and what's available.  There are some infrared cameras that can take 3D images to make torso meshes with electrode positions.  There is a company called Peacs that makes software for this.  Another way would be to make a geometry for the sensor positions based on the manufacture specifications and register it to the correct position.  SCIRun has some tools to register the sensor once you've created it in Matlab or Python.  Digitizing is another way.  Or you could try to estimate the positions.  I'm sure there are other ways too.

**Comment by zhifeiji-han** [comment]
Thank you for your reply！
If I just want to carry out software simulation , is there any way to obtain the sensor array position in the specified area?

**Comment by jessdtate** [comment]
you'll have to make the geometry for the sensor somehow (see previous comment for some ideas).  You can use map3d to add points and make triangles to the 771 tank geometry, but it isn't necessary for the sensors to be a part of the tank geom if you make the mapping matrix.  

If the sensor positions are a part of the torso geometry, and you know the index for them, you can extract the rows of the BE matrix that correspond to the node indices and that would be your transfer matrix.  This would make it so you don't need to make a geometry for the sensor positions.

**Comment by zhifeiji-han** [comment]
@jessdtate @jab0707 
I try to use map3d to add electrode points and make triangles to the 771 tank geometry.
I learned from the software manual that landmark can do it, as shown in the table below.
![image](https://user-images.githubusercontent.com/70640243/123720017-9a0aa080-d8b5-11eb-8714-d0a30761c6ed.png)
But I don't know how to implement it?

**Reply by zhifeiji-han** [reply]
@jessdtate
